source/kinsol_interface.cc

```diff
@@ -0,0 +1,426 @@
+
+    copy(b, *loc_b);
+
+    *sJpnorm = N_VWL2Norm(b, kin_mem->kin_fscale);
+    N_VProd(b, kin_mem->kin_fscale, b);
+    N_VProd(b, kin_mem->kin_fscale, b);
+    *sFdotJp = N_VDotProd(kin_mem->kin_fval, b);
+    return err;
+
+  }
+
+}
+
+template <typename VEC>
+void KINSOLInterface<VEC>::set_functions_to_trigger_an_assert()
+{
+
+  create_new_vector = []() ->shared_ptr<VEC>
+  {
+    shared_ptr<VEC> p;
+    AssertThrow(false, ExcPureFunctionCalled());
+    return p;
+  };
+
+  residual = [](const VEC &, VEC &) ->int
+  {
+    int ret=0;
+    AssertThrow(false, ExcPureFunctionCalled());
+    return ret;
+  };
+
+  setup_jacobian = [](const VEC &) ->int
+  {
+    int ret=0;
+    AssertThrow(false, ExcPureFunctionCalled());
+    return ret;
+  };
+
+  solve_linear_system = [](const VEC &, VEC &) ->int
+  {
+    int ret=0;
+    AssertThrow(false, ExcPureFunctionCalled());
+    return ret;
+  };
+
+  jacobian_vmult = [](const VEC &, VEC &) ->int
+  {
+    int ret =0;
+    AssertThrow(false, ExcPureFunctionCalled());
+    return ret;
+  };
+
+}
+
+// Class constructor:
+#ifdef DEAL_II_WITH_MPI
+
+
+template <typename VEC>
+KINSOLInterface<VEC>::KINSOLInterface(const std::string name,
+                                      const MPI_Comm &mpi_comm):
+  ParameterAcceptor(name),
+  is_initialized(false),
+  scaling_is_set(false),
+  communicator(mpi_comm),
+  pcout(std::cout, Utilities::MPI::this_mpi_process(mpi_comm)==0),
+  kin_mem(nullptr)
+{
+  set_functions_to_trigger_an_assert();
+}
+
+#else
+
+template <typename VEC>
+KINSOLInterface<VEC>::KINSOLInterface(const std::string name):
+  ParameterAcceptor(name),
+  is_initialized(false),
+  pcout(std::cout, Utilities::MPI::this_mpi_process(MPI_COMM_WORLD)==0),
+  kin_mem(nullptr)
+{
+  set_functions_to_trigger_an_assert();
+}
+
+#endif
+
+// Class destructor:
+template <typename VEC>
+KINSOLInterface<VEC>::~KINSOLInterface()
+{
+  if (kin_mem)
+    KINFree(&kin_mem);
+}
+
+// Parameters parsing and initialization of the parameters:
+template <typename VEC>
+void KINSOLInterface<VEC>::declare_parameters(ParameterHandler &prm)
+{
+
+  add_parameter(prm, &max_iterations,
+                "Maximum number of iterations","200",
+                Patterns::Integer(0),
+                "The non-linear solver will stop when reaching this number of"
+                "Newton iterations no matter what.");
+
+  add_parameter(prm, &ftol,
+                "Tolerance for residuals", "1e-6",
+                Patterns::Double(0),
+                "This define the condition (small residual) for a successful completion of KINSOL." );
+
+  add_parameter(prm, &steptol,
+                "Step tolerance", "1e-11",
+                Patterns::Double(0),
+                "The Newton method will terminate when the maximum scaled step is below the given tolerance." );
+
+  add_parameter(prm, &verbosity,
+                "Level of verbosity of the KINSOL solver", "0", Patterns::Integer(0),
```

Comment:
If the allowed values are 0,3, then change `Patterns::Integer(0)` to `Patterns::Integer(0,4)`.